File: 0189-rotate-array/0189-rotate-array.cpp
class Solution {
public:
    void reverse(vector<int>& nums, int i, int j)
    {
        while(i<=j)
        {
            int temp = nums[i];
            nums[i] = nums[j];
            nums[j] = temp;
            i++; j--;
        }
    }
    void rotate(vector<int>& nums, int k) {
        int n = nums.size();

        //the condition where k is greater than size of the vector length:
        while(k>n)
        {
            k = k-n;
        }

        int i = 0, j = n-1;
        reverse(nums, i, j);

        reverse(nums, 0, k-1);

        reverse(nums, k, n-1);
    }
};

File: 0189-rotate-array/0189-rotate-array.java
class Solution {
    public void rotate(int[] nums, int k) {
        int size = nums.length;
        // EDGECASE: WHEN k IS GREATER THAN THE LENGTH OF ARRAY 
        if (k > size) {
            k = k % size;
        }

        int start = 0;
        int end = size - 1;

        reverseArray(nums, start, end);
        
        end = k - 1;
        reverseArray(nums, start, end);
        
        start = k;
        end = size - 1;
        reverseArray(nums, start, end);

    }
    public static void reverseArray(int[] nums, int start, int end) {
        int size = nums.length;
        
        while (start < end) {
            int temp = nums[end];
            nums[end] = nums[start];
            nums[start] = temp;

            start++;
            end--;
        }
    }
}